mobile_verifier/src/heartbeats/mod.rs

```diff
@@ -282,165 +361,244 @@ impl ValidatedHeartbeat {
-                VALUES ($1, $2, $3, $4, $5, $6)
-                ON CONFLICT (hotspot_key, truncated_timestamp) DO UPDATE SET
-                latest_timestamp = EXCLUDED.latest_timestamp
-                RETURNING (xmax = 0) as inserted
-                "#,
+            INSERT INTO wifi_heartbeats (hotspot_key, cell_type, location_validation_timestamp, distance_to_asserted,
+                latest_timestamp, truncated_timestamp, coverage_object)
+            VALUES ($1, $2, $3, $4, $5, $6, $7)
+            ON CONFLICT (hotspot_key, truncated_timestamp) DO UPDATE SET
+            latest_timestamp = EXCLUDED.latest_timestamp,
+            coverage_object = EXCLUDED.coverage_object
+            "#,
         )
-        .bind(self.report.hotspot_key)
+        .bind(self.heartbeat.hotspot_key)
         .bind(self.cell_type)
-        .bind(self.report.location_validation_timestamp)
+        .bind(self.heartbeat.location_validation_timestamp)
         .bind(self.distance_to_asserted)
-        .bind(self.report.timestamp)
+        .bind(self.heartbeat.timestamp)
         .bind(truncated_timestamp)
+        .bind(self.heartbeat.coverage_object)
         .fetch_one(&mut *exec)
-        .await?
-        .inserted)
+        .await?;
+        Ok(())
     }
 }
 
 /// Validate a heartbeat in the given epoch.
 pub async fn validate_heartbeat(
     heartbeat: &Heartbeat,
-    gateway_client: &mut GatewayClient,
+    gateway_resolver: &impl GatewayResolver,
+    _coverage_cache: &CoveredHexCache,
     epoch: &Range<DateTime<Utc>>,
-) -> anyhow::Result<(CellType, proto::HeartbeatValidity, Option<i64>)> {
+    _max_distance: f64,
+) -> anyhow::Result<(
+    CellType,
+    Option<i64>,
+    Option<DateTime<Utc>>,
+    proto::HeartbeatValidity,
+)> {
     let cell_type = match heartbeat.hb_type {
-        HBType::Cbrs => match heartbeat.cbsd_id.as_ref() {
+        HbType::Cbrs => match heartbeat.cbsd_id.as_ref() {
             Some(cbsd_id) => match CellType::from_cbsd_id(cbsd_id) {
                 Some(ty) => ty,
                 _ => {
                     return Ok((
                         CellType::CellTypeNone,
-                        proto::HeartbeatValidity::BadCbsdId,
                         None,
+                        None,
+                        proto::HeartbeatValidity::BadCbsdId,
                     ))
                 }
             },
             None => {
                 return Ok((
                     CellType::CellTypeNone,
-                    proto::HeartbeatValidity::BadCbsdId,
                     None,
+                    None,
+                    proto::HeartbeatValidity::BadCbsdId,
                 ))
             }
         },
         // for wifi HBs temporary assume we have an indoor wifi spot
         // this will be better/properly handled when coverage reports are live
-        HBType::Wifi => CellType::NovaGenericWifiIndoor,
+        HbType::Wifi => CellType::NovaGenericWifiIndoor,
```

Comment:
whilst at this point we only have wifi indoor radios, still be good to update this so that the wifi type ( indoor/outdoor) is derived from the coverage object now that it is available

Reply:
I'll handle that in the phase 3 PR